paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc

```diff
@@ -0,0 +1,148 @@
+  if (!followed_eager_deletion_op) {
+    VLOG(4) << "No eager_deletion op found after this cinn_launch op";
+    return;
+  }
+  auto vars_to_delete = followed_eager_deletion_op->VarsToDelete();
+  if (vars_to_delete.empty()) {
+    VLOG(4) << "No var to be deleted after this cinn_launch op";
+    return;
+  }
+  VLOG(4) << "Variables would be deleted by eager_deletion_op"
+          << " following the cinn_launch:"
+          << paddle::string::join_strings(vars_to_delete, ',');
+
+  const auto& subgraph = paddle2cinn::CinnCompiler::GetInstance()->FindGraph(
+      cinn_launch_op->GetOp()->Attr<std::string>(operators::kCompilationKey));
+  auto& varinfo_from_maingraph =
+      subgraph.Get<Name2VarInfoMap>(paddle2cinn::kMemOptVarInfoFromMainGraph);
+  const auto& cur_place_var_infos = var_infos.at(cinn_launch_op->GetScopeIdx());
+
+  // collect all MemOptVarInfos of external variables
+  // that would be eager deleted after the cinn_launch subgraph executed,
+  // and store them as attribute of the subgraph
+  for (const auto& var_name : vars_to_delete) {
+    auto it = cur_place_var_infos.find(var_name);
+    PADDLE_ENFORCE_NE(it, cur_place_var_infos.end(),
+                      platform::errors::NotFound(
+                          "MemOptVarInfo of var[%s] not found", var_name));
+    varinfo_from_maingraph.emplace(var_name, it->second);
+  }
+}
+
+static void TakeVarInfoFromMainGraph(
+    const Name2VarInfoMap& parent_var_infos,
+    const MemOptVarInfoMapList& var_infos,
+    details::EagerDeletionOpHandle* eager_deletion_op) {
+  const auto& cur_place_var_infos =
+      var_infos.at(eager_deletion_op->GetScopeIdx());
+  for (auto&& var_name : eager_deletion_op->VarsToDelete()) {
+    auto cur_it = cur_place_var_infos.find(var_name);
+    PADDLE_ENFORCE_NE(cur_it, cur_place_var_infos.end(),
+                      platform::errors::NotFound(
+                          "MemOptVarInfo of var[%s] not found", var_name));
+    auto parent_it = parent_var_infos.find(var_name);
+    if (parent_it != parent_var_infos.end()) {
+      VLOG(4) << "MemOptVarInfo of var[" << var_name << "] set parent holder";
+      cur_it->second->SetParentHolder(parent_it->second);
+    }
+  }
+}
+
+// This pass will be applied on both the main graph and all subgraphs,
+// and it distinguishs them according to whether the graph has the
+// kMemOptVarInfoFromMainGraph attribute or not.
+// On the main graph, it finds all cinn_launch ops and shares MemOptVarInfos
+// to their subgraphs.
+// On a subgraph, it iterates each variable that will be deleted by a
+// eager_deletion op, and take the MemOptVarInfo from the main graph
+// if such one found.
+class ShareMemOptInfoToSubGraphPass : public ir::Pass {
+ protected:
+  void ApplyImpl(ir::Graph* graph) const override {
+    auto all_ops = ir::FilterByNodeWrapper<details::OpHandleBase>(*graph);
+    const auto& var_infos = Get<MemOptVarInfoMapList>(kMemOptVarInfoMapList);
+
+    // the main graph
```

Comment:
Is that possible a non-cinn subgraph can call a cinn subgraph?

Reply:
no, i update name of this pass and some comments to indicate clearly this pass will use with CINN